raft/storage.go

```diff
@@ -57,6 +57,17 @@ type Storage interface {
 	Snapshot() (pb.Snapshot, error)
 }
 
+// InitialState is the return value of the Storage.InitialState method.
+type InitialState struct {
+	HardState pb.HardState
+	ConfState pb.ConfState
+
+	// AppliedIndex is the last index which has been applied to the state machine,
+	// for implementations which persist the state machine to disk. If the state
+	// machine is not persistent, this field may be zero.
+	AppliedIndex uint64
```

Comment:
We are going to have a disk backend applier soon. 
The workflow is:
1. raft commits entries
2. raft sends out entries to etcd.raft
3 etcd.raft passes entries to applier
3 etcd.raft.Storage saves the entries to disk (concurrently)
4. applier applies entries and update applied index
5. applier sends back ack to etcd.raft
6. etcd.raft waits for the ack from applier and etcd.raft.Storage.Save, then advances raft

Since etcd.raft.Storage and applier will be working on entries concurrently. We cannot reliably save appliedIndex into Storage.